### PR TITLE
[CircleEditor] Preprocessing for Editing Circle Files

### PR DESCRIPTION
**Pull request by Seonhyuk**
This PR introduces preprocessing for editing circle file. It enables saving the data from open window view, adding data necessary for data modification, and adding vscode cmds and types of custom operator attributes.

ONE-vscode-DCO-1.0-Signed-off-by: Seonhyuk Ahn <gomduridkstjsgur@gmail.com>

---

Related issue : #1269
Draft : #1298

### CODE REVIEW

File: media/CircleEditor/index.js [hunk 1]
```diff
@@ -146,6 +150,12 @@ host.BrowserHost = class {
                 case 'reload':
                     this._msgReload(message);
                     break;
+                case 'edit':
+                    this._msgLoadModel(message);
+                    break;
+                case 'customType':
+                    this._msgGetType(message);
```
Comment (llFreetimell) [inline media/CircleEditor/index.js:157]: It's hard to think about the meaning of this function :(
What is this function for?

File: media/CircleEditor/view.js [hunk 2]
```diff
@@ -821,6 +825,24 @@ view.View = class {
                     if (this._page !== 'default') {
                         this.show('default');
                     }
+                    for (let idx = 0; idx < model.graphs.length; idx++) {
+                        model._graphs[idx]['_subgraphIdx'] = idx;
+                        for (let jdx = 0; jdx < model.graphs[idx].nodes.length; jdx++) {
+                            model._graphs[idx]._nodes[jdx]['_subgraphIdx'] = idx;
+                            if (builtinOperatorType
+                                    [model._graphs[idx]._nodes[jdx]._type.name.toUpperCase()] ===
```
Comment (llFreetimell) [inline media/CircleEditor/view.js:833]: As far as I know, `CUSTOM` is defined in `builtinOperatorType` as 32.
Isn't it error when `builtinOperatorType` is undefined?

File: media/CircleEditor/view.js [hunk 3]
```diff
@@ -168,6 +168,10 @@ view.View = class {
             }
         }
         this._page = page;
+
+        if (typeof (this._host._viewingNode) === 'number') {
```
Comment (llFreetimell) [inline media/CircleEditor/view.js:172]: How about following?
At first, I misunderstood that `_viewingNode` can be another type than `number` :)

```suggestion
        if (this._host._viewingNode !== null) {
```

File: media/CircleEditor/view.js (hunk 2 above)
Comment (llFreetimell) [reply media/CircleEditor/view.js:833]: And this is just optional, but it would be better to introduce a function like `builtinOperatorTypeAsNumber(string)` :)

File: media/CircleEditor/index.js [hunk 4]
```diff
@@ -46,7 +46,8 @@
 
 var host = {};
 
-const vscode = acquireVsCodeApi();
+var vscode = acquireVsCodeApi();
```
Comment (llFreetimell) [inline media/CircleEditor/index.js:49]:
```suggestion
let vscode = acquireVsCodeApi();
```

File: media/CircleEditor/index.js (hunk 1 above)
Comment (Seonhyuk) [reply media/CircleEditor/index.js:157]: For attributes of the custom operator, the type is not recognized by netron. So it's a function that takes the type from the model and fills in the attributes type of the custom operator.

File: media/CircleEditor/view.js (hunk 3 above)
Comment (Seonhyuk) [reply media/CircleEditor/view.js:172]: Thank you! I think it's a good idea!

File: media/CircleEditor/index.js (hunk 4 above)
Comment (Seonhyuk) [reply media/CircleEditor/index.js:49]: It doesn't have to be declared again. I will use `const` as it is. Thank you!

File: media/CircleEditor/view.js (hunk 2 above)
Comment (Seonhyuk) [reply media/CircleEditor/view.js:833]: I was found a custom operator classification method in the netron model. I will use the if statement as like `model._graphs[idx]._nodes[jdx]._type?.category === 'custom'`. Thank you!!!

File: media/CircleEditor/index.js (hunk 4 above)
Comment (llFreetimell) [reply media/CircleEditor/index.js]: If `acquireVsCodeApi()` is called twice, it would cause an error.
The solution is creating a file like `media/CfgEditor/vscodeapi.js` and importing it :)

File: media/CircleEditor/index.js (hunk 1 above)
Comment (llFreetimell) [reply media/CircleEditor/index.js:156]: As there are so many kinds of "type", it is really confused :(
For example,
- `customType` : Is there any type like this? Which type in custom operator, attributes' or inputs' or outputs'?
- `msgGetType` : Which type can be gotten? All operators or only some of operators?

Could you clarify the names?

File: media/CircleEditor/index.js [hunk 5]
```diff
@@ -146,6 +149,12 @@ host.BrowserHost = class {
                 case 'reload':
                     this._msgReload(message);
                     break;
+                case 'edit':
+                    this._msgLoadModel(message);
```
Comment (llFreetimell) [inline media/CircleEditor/index.js:153]: Only seeing the `command`, `edit` seems not related with loading model.
Furthermore, this case is exactly same with `loadmodel`.
What is this case for?

File: media/CircleEditor/index.js [hunk 6]
```diff
@@ -537,10 +546,19 @@ host.BrowserHost = class {
...
     }
 
-    /**
-     * TODO Implement _msgGetType(message)
-     * Get custom operator attributes type
-     */
+    _msgGetType(message) {
+        const data = message.data;
+        const graphs = data._view._model._graphs;
+        const values = data._type;
```
Comment (llFreetimell) [inline media/CircleEditor/index.js:552]: For consistency, following seems better :)

```suggestion
        const types = data._type;
```

File: media/CircleEditor/index.js (hunk 5 above)
Comment (Seonhyuk) [reply media/CircleEditor/index.js:153]: I separated the `edit` command considering the possibility of partial modification in the future, but I think it can be combined into `loadmodel` :)

File: media/CircleEditor/index.js (hunk 6 above)
Comment (Seonhyuk) [reply media/CircleEditor/index.js:552]: Thank you!

File: media/CircleEditor/index.js (hunk 1 above)
Comment (Seonhyuk) [reply media/CircleEditor/index.js:153]: There could be some confusion... `msgGetCustomOperatorAttributesType` I think it will be a little long to express them all clearly, do you have any good ideas?

Comment (llFreetimell) [reply media/CircleEditor/index.js:153]: Hmm... how about `msgGetCustomOpAttrT`?
It's little bit longer too but shorter than full name...

Comment (Seonhyuk) [reply media/CircleEditor/index.js:153]: I'd like that. I'll apply it! Thank you!

File: media/CircleEditor/view.js [hunk 7]
```diff
@@ -821,6 +825,22 @@ view.View = class {
...
                         this.show('default');
                     }
+                    for (let idx = 0; idx < model.graphs.length; idx++) {
+                        model._graphs[idx]['_subgraphIdx'] = idx;
+                        for (let jdx = 0; jdx < model.graphs[idx].nodes.length; jdx++) {
+                            model._graphs[idx]._nodes[jdx]['_subgraphIdx'] = idx;
+                            if (model._graphs[idx]._nodes[jdx]._type?.category === 'custom') {
+                                model._graphs[idx]._nodes[jdx]._isCustom = true;
+                                vscode.postMessage({
+                                    command: 'getCustomOpAttrT',
```
Comment (llFreetimell) [inline media/CircleEditor/view.js:835]: I am little bit confused...
It seems that there are two `getCustomOpAttrT`, which are in `index.js` and `CircleEditorProvider.ts`.
Then the two message listeners are really for **getting** custom operator attributes type? not **setting**?
Furthermore, there is no `CircleEditorProvider.ts` to handle this message...

Comment (Seonhyuk) [reply media/CircleEditor/view.js:835]: I'll change `get-` to `set-` in `index.js`. 
Would it be better to upload the related parts of the message after `CircleEditorProvider.ts` is uploaded? 
I think that sending and receiving messages is always complementary, so what should we do with the commands that already exist?

Comment (llFreetimell) [reply media/CircleEditor/view.js:835]: > Would it be better to upload the related parts of the message after CircleEditorProvider.ts is uploaded?

Leaving it as TODO and then implementing after message listeners in `CircleEditorProvider.ts` is introduced would be better :)

> so what should we do with the commands that already exist?

What does this mean...? T.T

Comment (Seonhyuk) [reply media/CircleEditor/view.js]: I left the related part as TODO..

**Review by llFreetimell — APPROVED** [review]
LGTM :)

**Review by YuyeonHan — APPROVED** [review]
LGTM :)
